### PR TITLE
Fix race condition introduced by background platform channels

### PR DESCRIPTION
**Pull request by blasten**
Fixes race condition when the Dart VM starts before platform channel handlers registration. 
This is the case when background platform channels are used because there's no locking mechanism 
between the Dart UI thread and the platform thread.

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 1]
```diff
@@ -188,39 +269,65 @@ private void invokeHandler(
...
 
+  @Override
+  public void handleMessageFromDart(
+      @NonNull String channel, @Nullable ByteBuffer message, int replyId, long messageData) {
+    // Called from the ui thread.
+    Log.v(TAG, "Received message from Dart over channel '" + channel + "'");
+
+    HandlerInfo handlerInfo;
+    synchronized (handlersLock) {
+      if (messageHandlers.contains(channel)) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:308]: If the message was not deferred, then it must be dispatched even if the handlerInfo is null.

Try something like:
```
synchronized (handlersLock) {
  handlerInfo = messageHandlers.get(channel)
  messageDeferred = (canDelayTasks && handlerInfo == null);
  if (messageDeferred) {
    // add it to the deferred message list
  }
}
if (!messageDeferred) {
  // process the message now
}

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 2]
```diff
@@ -188,39 +269,65 @@ private void invokeHandler(
...
+        //
+        // This is possible if the Dart VM starts before channel registration, and if the thread
+        // that registers the channels is busy or slow at registering the channel handlers.
+        //
+        // In such cases, the task dispatchers are queued, and processed when the channel is
+        // defined.
+        if (!delayedTaskDispatcher.contains(channel)) {
+          delayedTaskDispatcher.put(channel, new LinkedList<>());
+        }
+        List<Runnable> delayedTaskQueue = delayedTaskDispatcher.get(channel);
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:322]: Rename this so it is not mistaken for the `TaskQueue` interface that is used to execute message handlers

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 3]
```diff
@@ -188,39 +269,65 @@ private void invokeHandler(
...
             invokeHandler(handlerInfo, message, replyId);
             if (message != null && message.isDirect()) {
-              // This ensures that if a user retains an instance to the ByteBuffer and it happens to
-              // be direct they will get a deterministic error.
+              // This ensures that if a user retains an instance to the ByteBuffer and it
+              // happens to be direct they will get a deterministic error.
               message.limit(0);
             }
           } finally {
+            Trace.endSection();
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:290]: Move this below `flutterJNI.cleanupMessageData`

`flutterJNI.cleanupMessageData` must be called in order to avoid memory leaks.  Calling it first avoids the potential risk of `Trace.endSection` throwing.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 4]
```diff
@@ -124,19 +165,59 @@ public void setMessageHandler(
...
+
+  /**
+   * Runs the tasks that handle messages received from Dart for the provided channel name.
+   *
+   * <p>The channel may not have associated tasks if it was registered prior to reciving the first
+   * message from Dart.
+   *
+   * @param channel The channel name.
+   */
+  public void runDelayedTasksForChannel(@NonNull String channel) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:196]: Merge this into `setMessageHandler`

The `synchronized (handlersLock)` block in `setMessageHandler` can call `delayedTaskDispatcher.remove(channel)` and retain the return value to get the deferred message list.

`setMessageHandler` already has the `HandlerInfo`, so the loop that processes the deferred message list can then avoid the `messageHandlers.get()` call

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 5]
```diff
@@ -124,19 +165,59 @@ public void setMessageHandler(
...
+    LinkedList<DelayedMessageInfo> list;
+    synchronized (handlersLock) {
+      if (!delayedTaskDispatcher.contains(channel)) {
+        return;
+      }
+      list = (LinkedList) delayedTaskDispatcher.get(channel);
+      delayedTaskDispatcher.remove(channel);
+    }
+    while (!list.isEmpty()) {
+      final DelayedMessageInfo info = list.poll();
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:206]: No need to call `poll` and update the list.  Just iterate over the list and it will then be GCed.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 6]
```diff
@@ -33,22 +35,41 @@
...
 
-  @NonNull private final ConcurrentHashMap<String, HandlerInfo> messageHandlers;
+  /**
+   * Maps a channel name to an object that contains the task queue and the handler associated with
+   * the channel.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull
+  private final ConcurrentHashMap<String, HandlerInfo> messageHandlers = new ConcurrentHashMap<>();
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:45]: This no longer needs to be a `ConcurrentHashMap` now that all accesses are guarded by a lock.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 7]
```diff
@@ -33,22 +35,41 @@
...
+
+  /**
+   * Maps a channel name to queue of task dispatchers. This queue is processed when the channel
+   * handler is registered.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull
+  private final ConcurrentHashMap<String, List<DelayedMessageInfo>> delayedTaskDispatcher =
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:54]: Update the name and comment now that this no longer holds task dispatchers.

This can also be a non-concurrent `HashMap`

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 8]
```diff
@@ -124,19 +165,59 @@ public void setMessageHandler(
...
+  }
+
+  /**
+   * Enables the ability to queue tasks when messages are received from Dart.
+   *
+   * <p>This is useful when there are pending channel handler registrations. For example, Dart may
+   * be initialized concurrently, and prior to the registration of the channel handlers. This
+   * implies that Dart may start sending messages while plugins are being registered.
+   */
+  public void enableDelayedTaskQueue() {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:219]: I'd prefer a different name to avoid reusing the term `TaskQueue`

Suggestion: `enableBufferingIncomingMessages`?

If this is renamed, the `canDelayTasks` flag name should also be changed to match

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 1 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: done

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 9]
```diff
@@ -33,22 +34,37 @@
...
   @NonNull private final FlutterJNI flutterJNI;
 
-  @NonNull private final ConcurrentHashMap<String, HandlerInfo> messageHandlers;
+  /**
+   * Maps a channel name to an object that contains the task queue and the handler associated with
+   * the channel.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull private final Map<String, HandlerInfo> messageHandlers = new HashMap<>();
```
Comment (dnfield) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:43]: Instead of two `Map` objects, could we jus tkeep one `ConcurrentHashMap` and either create a wrapper object for the buffered messages or add them to `HandlerInfo`?

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 10]
```diff
@@ -215,12 +273,43 @@ public void handleMessageFromDart(
...
+      messageDeferred = (enableBufferingIncomingMessages && handlerInfo == null);
+      if (messageDeferred) {
+        // The channel is not defined when the Dart VM sends a message before the channels are
+        // registered.
+        //
+        // This is possible if the Dart VM starts before channel registration, and if the thread
+        // that registers the channels is busy or slow at registering the channel handlers.
+        //
+        // In such cases, the task dispatchers are queued, and processed when the channel is
+        // defined.
```
Comment (dnfield) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:300]: Can't this also happen if the handler gets unregistered?

File: shell/platform/android/test/io/flutter/embedding/engine/dart/DartMessengerTest.java [hunk 11]
```diff
@@ -173,17 +179,63 @@ public void cleansUpMessageData() throws InterruptedException {
...
+    messenger.setMessageHandler(channel, handler, taskQueue);
+
+    shadowOf(getMainLooper()).idle();
+    verify(fakeFlutterJni, never()).invokePlatformMessageEmptyResponseCallback(eq(replyId));
+
+    final ArgumentCaptor<ByteBuffer> response = ArgumentCaptor.forClass(ByteBuffer.class);
+    verify(fakeFlutterJni)
+        .invokePlatformMessageResponseCallback(anyInt(), response.capture(), anyInt());
+    assertArrayEquals("done".getBytes(), response.getValue().array());
+  }
```
Comment (dnfield) [inline shell/platform/android/test/io/flutter/embedding/engine/dart/DartMessengerTest.java:240]: Can you add a test that handles when a handler gets _unset_ and a message is sent from Dart?

I'm not entirely clear on how we should handle that, but it seems like that should end up throwing a MissingPluginException. But maybe it's not a case we're too concerned about, I can't think of why this would happen.

A case that _is_ a bit more concerning is someone forgets to ever register the plugin but still tries to send a message on it. E.g. when developing a plugin you have a bug you can't really discover easily.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 12]
```diff
@@ -124,21 +160,47 @@ public void setMessageHandler(
...
+    LinkedList<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
+        return;
+      }
+      list = (LinkedList) bufferedMessages.get(channel);
+      bufferedMessages.remove(channel);
+    }
+    for (BufferedMessageInfo info : list) {
```
Comment (gaaclarke) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: This isn't thread-safe.  The ui thread could be adding to the linked list while the platform thread is iterating over it.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 13]
```diff
@@ -215,12 +273,43 @@ public void handleMessageFromDart(
...
+        // The channel is not defined when the Dart VM sends a message before the channels are
+        // registered.
+        //
+        // This is possible if the Dart VM starts before channel registration, and if the thread
+        // that registers the channels is busy or slow at registering the channel handlers.
+        //
+        // In such cases, the task dispatchers are queued, and processed when the channel is
+        // defined.
+        if (!bufferedMessages.containsKey(channel)) {
+          bufferedMessages.put(channel, new LinkedList<>());
```
Comment (gaaclarke) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:302]: When I implemented Channel Buffers I got a lot of flak about having unbounded buffering of platform channels.  The resolution we ultimately decided upon was a RingBuffer of size 1 with control messages that can be used to adjust the size.  The concern was that incorrect usage could blow up memory.

I thought it was overly cautious.  Maybe we should have a RingBuffer at least so it isn't completely unbounded.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: That case would be handled by the normal flow. Since L180 already populated messageHandlers, there should be no issue AFAIT

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 13 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:302]: is the suggestion to override messages? How is that desired in an app that runs into this situation?

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: The problem isn't `messageHandlers`, the LinkedList data structure can be simultaneously written to and read from 2 different threads at the same time.  Here it is being read on L187 unguarded and it can be written to on L305.  Reading and writing to a non-threadsafe data structure simultaneously can result in unpredictable results.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 9 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:43]: The wrapper object would still need to represent two different states. One with a Handler, and one without a handler, but with buffered messages.  FWIW, having two separate maps provides a more transparent separation of concerns.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: `synchronized(lock)` is handling the locking. Is there an issue with this lock?

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 10 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:300]: that is right. In this case, `enableBufferingIncomingMessages` should be set to `false`. That makes sense to add.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 13 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:302]: A RingBuffer drops messages after its capacity is exhausted, dropping messages isn't a desired behavior but also queueing up infinite messages isn't desired.

Messages that write data can have 2 different semantics, `x = y` or `x = f(x, y)`.  In the first case dropping messages isn't actually an error.  The only thing that matters is the last operation.  In the latter case, it is a problem.  It impossible to know which write semantics the channel has without the user declaring it.

So, in summary, a RingBuffer doesn't always drop messages and when it does it isn't necessarily a logical error.  If I were king of Flutter I'd say just chose some reasonable upper bounds for the RingBuffer, like 1024.  I can tell you though that we've had this discussion before as a team and the resolution was that the RingBuffer size should be 1.

attn @Hixie who was involved in deciding how Channel Buffers should work.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: Yes, that's what I'm telling you =)   Specifically there is no lock while you are reading from the LinkedList, but there is one when you are writing to it.  When you read the List from bufferedMessages it is passed by reference, not by value, it's the same datastructure being written to and read from, only the write has a lock.

Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:187]: My understanding of this problem is:
1. Once we exited the synchronization block above, `messageHandlers` has been set. At this point, a thread trying to get the lock in the second synchronization block runs.
2. The second synchronization block realizes that the `messageHandler` is in the map, therefore `messageDeferred` returns `false`, and no write to the buffer list occurs. 

As a result, there's no case in which this loop is running and a write occurs in the middle.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 13 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:302]: If a RingBuffer is preferred, I'm happy to change it. I wasn't aware that it's ok to drop some messages. e.g. could that yield to invalid state in a way that is really hard to debug?

File: shell/platform/android/test/io/flutter/embedding/engine/dart/DartMessengerTest.java (hunk 11 above)
Comment (blasten) [reply shell/platform/android/test/io/flutter/embedding/engine/dart/DartMessengerTest.java:245]: Done

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:216]: Right, because of the removal and acquisition of the list are guarded, this usage is correct.  What I've said above is true but isn't a problem in practice.  There is no guarantee that someone isn't accessing the LinkedList while we are reading from it, but we don't do it.  I'm trying to think if there is any change I recommend, a comment or something.

If `bufferedMessages` was a ConcurrentMap and the LinkedList implementation was persistent you could guarantee correctness (unfortunately there isn't a persistent linked list implementation in the sdk).  Did you not want to use ConcurrentMap because you didn't want the overhead of locking twice?  That's probably the right call.

This is fine with me, I can't think of a comment or easy change that makes this better given what the sdk provides.  Thanks for walking through it.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 13 above)
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:340]: Is the plan that buffering messages only happens for a limited amount of time via `enableBufferingIncomingMessages`?  It doesn't appear to be doing that but if that's the case it isn't as bad since you only have a limited amount of time to queue up infinite messages.

What is the plan for messages that are sent, buffered, but no handler is ever registered for them?  Are you going to hold onto them indefinitely?  In the past they would have gotten an exception that maybe was ignored but now will be held onto.  Maybe once we turn off the time frame where messages are buffered all of those messages without handlers could be cleared out to match the previous behavior.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 12 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:216]: right. Do you know if `ConcurrentMap` is actually preventing races? This part from the docs is confusing:

> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove)

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:216]: Yea, it's thread-safe.  That just means it is implemented in such a way that you won't have the overhead of a lock when doing read operations which is cool, that's what we'd want.  Using a lock, reading and writing would be serialized.  It also means that technically a read and a write at the same exact time might not strictly happen in order.  That's why it has methods that read and write the value atomically (like merge).

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 14]
```diff
@@ -154,21 +189,56 @@ public void setMessageHandler(
...
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+
+    LinkedList<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
+        return;
+      }
+      list = (LinkedList) bufferedMessages.get(channel);
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:213]: Remove the `get` and just call `remove`.  `Map.remove` will return the removed value (if any)

Also change `list` to a `List<BufferedMessageInfo>` to match the value type of `bufferedMessages`.  And this cast should use the `List<BufferedMessageInfo>` parameterized type.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 15]
```diff
@@ -33,22 +34,37 @@
...
+  /**
+   * Maps a channel name to an object that holds information about the incoming Dart message.
+   *
+   * <p>Reads and writes to this map must lock {@code handlersLock}.
+   */
+  @NonNull private final Map<String, List<BufferedMessageInfo>> bufferedMessages = new HashMap<>();
 
-  @NonNull private final Map<Integer, BinaryMessenger.BinaryReply> pendingReplies;
+  @NonNull private final Object handlersLock = new Object();
+  private boolean enableBufferingIncomingMessages = false;
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:53]: Make this an `AtomicBoolean`
(it will be set on the platform thread and read on the UI thread)

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 16]
```diff
@@ -154,21 +189,56 @@ public void setMessageHandler(
...
   }
 
+  /**
+   * Enables the ability to queue messages received from Dart.
+   *
+   * <p>This is useful when there are pending channel handler registrations. For example, Dart may
+   * be initialized concurrently, and prior to the registration of the channel handlers. This
+   * implies that Dart may start sending messages while plugins are being registered.
+   */
+  public void enableBufferingIncomingMessages() {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:229]: This should be added to the public `BinaryMessenger` API

(`DartMessenger` is not a public class)

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 13 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:341]: > Is the plan that buffering messages only happens for a limited amount of time via enableBufferingIncomingMessages?

right on

> What is the plan for messages that are sent, buffered, but no handler is ever registered for them?

Added some logic to address this in disableBufferingIncomingMessages

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 14 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: nice

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 15 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java]: right on

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java (hunk 16 above)
Comment (blasten) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:222]: good catch

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 17]
```diff
@@ -154,18 +189,54 @@ public void setMessageHandler(
...
+        throw new IllegalArgumentException(
+            "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
+      }
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+
+    List<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:210]: This lookup can be avoided by just calling `remove` and then checking whether `remove` returned null

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 18]
```diff
@@ -154,18 +189,54 @@ public void setMessageHandler(
...
+  }
+
+  @Override
+  public void disableBufferingIncomingMessages() {
+    if (!enableBufferingIncomingMessages.getAndSet(false)) {
+      return;
+    }
+    Map<String, List<BufferedMessageInfo>> pendingMessagesCopy = new HashMap<>();
+    synchronized (handlersLock) {
+      pendingMessagesCopy = new HashMap(bufferedMessages);
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:233]: This copy can be avoided by setting `pendingMessages` to the current value of `bufferedMessages` and then setting `bufferedMessages` to a newly allocated map.

This is safe because `bufferedMessages` is only accessed while holding the `handlersLock`.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 19]
```diff
@@ -154,18 +189,54 @@ public void setMessageHandler(
...
+      bufferedMessages.clear();
+    }
+    for (String channel : pendingMessagesCopy.keySet()) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:236]: Iterate over the map's `entrySet` in order to avoid doing the key lookups

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 20]
```diff
@@ -154,18 +189,54 @@ public void setMessageHandler(
...
+
+  @Override
+  public void enableBufferingIncomingMessages() {
+    enableBufferingIncomingMessages.set(true);
+  }
+
+  @Override
+  public void disableBufferingIncomingMessages() {
+    if (!enableBufferingIncomingMessages.getAndSet(false)) {
+      return;
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:229]: There is a potential race if you return here.  `handleMessageFromDart` could read `enableBufferingIncomingMessages` as true immediately before this write and then proceed to insert something into `bufferedMessages`.

It's safer to skip this return and always obtain the `handlersLock` and drain anything that may have been buffered.

**Review by jason-simmons — APPROVED** [review]

File: shell/platform/android/io/flutter/plugin/common/BinaryMessenger.java [hunk 21]
```diff
@@ -116,6 +116,22 @@ default void setMessageHandler(
...
   }
 
+  /**
+   * Enables the ability to queue messages received from Dart.
+   *
+   * <p>This is useful when there are pending channel handler registrations. For example, Dart may
+   * be initialized concurrently, and prior to the registration of the channel handlers. This
+   * implies that Dart may start sending messages while plugins are being registered.
+   */
+  void enableBufferingIncomingMessages();
```
Comment (gaaclarke) [inline shell/platform/android/io/flutter/plugin/common/BinaryMessenger.java:126]: There going to be another PR that will enable and disable the buffering during the timeframe of plugin registration I imagine?

Comment (blasten) [reply shell/platform/android/io/flutter/plugin/common/BinaryMessenger.java:126]: that's right

**Comment by blasten** [comment]
The MacOS built succeeded after a retry, but for some reason it's not propagating. 
https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20Unopt/5680/overview

cc @godofredoc
